### ISSUE TITLE
Improve Tutorials / Samples

### ISSUE THREAD

**Issue by geoeo** [mellinoe/veldrid#30]
It would be really nice to have a more complete set of tutorials/samples to work with.

E.g. None of the samples use Geometry Instancing/Indirect Buffers.
From the Vulkan/Metal docs it is also not clear, because each API handles this slightly differently. 

Best,
Marc

**Comment by mellinoe** [comment]
Yeah I agree, more example code would be nice. I've gotten requests for a lot of different types of examples, so it's hard to choose which ones to implement in the time I have for it.

**Comment by geoeo** [comment]
Sure. Thats totally understandable

For me at least I like to reimplement stuff from  [learnOpenGL](https://learnopengl.com/)
I think the advanced examples are very good.

**Comment by Krakean** [comment]
@mellinoe 
By the way, do you accept donates? I would like to donate for tutorials/samples purposes.
For example, would be nice to have some of these tutorials http://www.rastertek.com/tutdx11.html or http://www.rastertek.com/tutdx11s2ter.html to be adapted (copy-paste is welcomed too, in context of technique descriptions, since these tutorials pretty well written though a lot of code) to Veldrid & C#.

**Comment by mellinoe** [comment]
@Krakean I don't take donations at the moment, no. For me, it's just a matter of not having enough time to dedicate to samples and all of my other projects. It's still something I want to do, though, and I understand they are very important for new folks trying to learn the library.

Which of the samples from rastertek do you think would be the most interesting to port over to Veldrid?

**Comment by geoeo** [comment]
For me I like to focus on these topics when I start with a new API:

- Uniforms
- Instanced Buffers
- Offscreen Rendering

Granted Veldrid-Samples covers these aspects, but they are a bit convoluted.
The biggest offender is Shader-Gen (sorry :octopus: ). 
I think its really cool but it doesn't help when wanting to do the absolute minimum in terms of shader programming.

I think a good first step would be to rewrite the shaders in their simplest form.

**Comment by feliwir** [comment]
@geoeo i think shader gen really helps with doings the minimum work required... Not using shader gen is 4 times the work

**Comment by geoeo** [comment]
@feliwir I dont think it helps when trying to learn a new Graphics API. Since its 2 technologies competing in the same tutorial. I try to keep things separated when I'm learning new things. 

But thats just how I like to learn stuff.

**Comment by mellinoe** [comment]
Writing the shaders 4 times is obviously more work, but I can understand where the feedback is coming from. I've tried to keep the "basic" tutorials off of ShaderGen so that they are more approachable. Some of the others were actually ported over from other code of mine which was already using ShaderGen. I think I agree with the general sentiment that the introductory example code should try to be as self-contained and simple as possible, even if it means writing the shaders multiple times. Honestly, it's not THAT much work for me to copy-paste the shaders and adapt them to different languages.

**Comment by vpenades** [comment]
If there's a poll for examples... my vote goes on skinning/morphing...  I am working on a generic c# glTF loader, and I was considering using veldrid as a render engine as a proof of concept.

In terms of animation, something that keeps me puzzled is how to achieve morphed animation using the GPU... theoretically a model can come with many morph targets, but the whole thing is not very GPU friendly... at least to my rusted GPU knowledge

**Comment by mellinoe** [comment]
@vpenades I think it would be great to use Veldrid for that. I'm happy to provide pointers if something is unclear.

There is an existing sample that covers (basic) skinning, but nothing with morph targets. I'm not really familiar with morphing animations (seems like they aren't widely used in games), but the naive way to do it would be to simply upload all of the vertex keyframe data into a large buffer. Each frame, upload a new time value and do the keyframe interpolation in the vertex shader.

**Comment by vpenades** [comment]
@mellinoe Morph targets are quite tricky to handle, and with glTF they're specially complex, because they can be used along with skinning animation. They're typically used for things like facial animation or cloth.

glTF defines a number of vertex channels, the classic ones: Position, Normal, Texcoords, skin weights, etc. Any of these channels can be subject to morphing, so it can lead to lots of weird combinations.

In practice, what I've seen is morph targets that have Position, Normal and Tangent for every channel.

But with glTF, any attribute can be morphed; for example, Texture coordinates, to make texture distortion animation over a surface. or even skin weights themselves, to change the weights based on pose.

There's also the issue of the limit of attributes a vertex can have, which, I think, it's 16.  If you have 3 morph targets binded, which have Position+Normal+Tangent, they already consume 9 out of 16 slots, leaving 7 for skinning, texture coords, color, etc.... it's really at the limit.

**Comment by mellinoe** [comment]
I guess I was thinking of it in a slightly different way. Instead of passing the data as per-vertex attributes, you could just pass them in a buffer ([ResourceKind.StructuredBufferReadOnly](https://mellinoe.github.io/veldrid-docs/articles/shaders.html#structured-buffer)), and read them in the vertex shader. This is sometimes called "vertex pulling". In my mind, you need to do this anyways because you'll want to sample multiple keyframes in the shader to interpolate between them. But again, I don't really know much about this kind of morph animation, and I'm getting the impression that glTF supports a lot of complex features that might require something different

> If you have 3 morph targets binded, which have Position+Normal+Tangent, they already consume 9 out of 16 slots

I think most systems nowadays support 32 attributes, but I think you will be fine if some (or all) of the animated attributes end up in a structured buffer.

**Comment by vpenades** [comment]
you're probably right; my GPU knowledge stops at DirectX9, things like structured buffers, etc are unknown to me.

my gltf repo is [here](https://github.com/vpenades/Epsylon.glTF2Toolkit), it's still a mess and I'm still figuring out namespaces and which helper classes I include or leave them out.  The interesting bit is the SharpDX demo, which I use as an use case. It just loads the geometry into Dx and renders it.

**Comment by opekope2** [comment]
I'd like to use Veldrid with GTK# but I can't find anything like that. I also would like to know how to get back the texture from the GPU and export it to a stream as PNG for example
